base/src/instantiate.rs

```diff
@@ -196,24 +196,62 @@ pub fn instantiate<F>(typ: TcType, mut f: F) -> TcType
                               })
 }
 
-
+/// Removes layers of `Type::App` and `Type::Data`.
+///
+/// Example:
+///
+/// ```
+/// use gluon_base::types::{Type, TcType, BuiltinType};
+/// use gluon_base::instantiate::unroll_app;
+/// let i: TcType = Type::int();
+/// let s: TcType = Type::string();
+/// assert_eq!(unroll_app(&*Type::app(Type::app(i.clone(), s.clone()), i.clone())),
+///            Some(Type::data(i.clone(), vec![s.clone(), i.clone()])));
+/// assert_eq!(unroll_app(&*Type::data(Type::app(i.clone(), i.clone()), vec![s.clone()])),
+///            Some(Type::data(i.clone(), vec![i.clone(), s.clone()])));
+/// let f: TcType = Type::builtin(BuiltinType::Function);
+/// assert_eq!(unroll_app(&*Type::data(Type::app(f.clone(), i.clone()), vec![s.clone()])),
+///            Some(Type::function(vec![i.clone()], s.clone())));
+/// ```
 pub fn unroll_app(typ: &Type<Symbol>) -> Option<TcType> {
     let mut args = Vec::new();
-    let mut current = typ;
+    let mut current = match *typ {
+        Type::App(ref l, ref r) => {
+            args.push(r.clone());
+            l
+        }
+        Type::Data(ref l, ref rest) => {
+            args.extend(rest.iter().rev().cloned());
+            l
+        }
+        _ => return None,
+    };
     loop {
-        match *current {
+        match **current {
             Type::App(ref l, ref r) => {
                 args.push(r.clone());
-                current = &**l;
+                current = l;
```

Comment:
Those dereferences used to be necessary actually. In the commit I also changed the type of `current` to be `TcType` instead of `Type<Symbol>` (TcType derefs to `Type<Symbol>`). I suppose it might have been cleaner to declare the type of `current` to get automatic dereferences before though.